### ISSUE TITLE
VideoContext state description error in api

### ISSUE THREAD

**Issue by Statfine** [bbc/VideoContext#25]
http://bbc.github.io/VideoContext/api/VideoContext.html 
<VideoContext.STATE.PLAYING: all sources are active>; I think the ‘all sources’ maybe change  to ‘current sources’ is more precise;
because i set a right url from 0s to 10s;and then set a error url from 15s to 20s; the VideoContext.state is STALLED after play 15s; so i think it is the current source state.
other i want to know is there some mark or status let me know from 10s to 15s VideoContext not has sources it‘s empty content

**Comment by MatthewShotton** [comment]
Hi,
Good spot! I think you're correct, "current sources" is a better description. I've updated the documentation appropriately.

Regarding checking the state of the videocontext for upcoming sources: This is a bit tricky, as the VideoContext doesn't try loading stuff until about 4 seconds until it is to be played. This is to prevent a whole load of requests happening right at the point play() is called which is bad because some browsers limit the number of simultaneous requests allowed to a single domain, and because the requests the VideoContext makes are typically for large video files which take a while to resolve this can saturate all available connections to a domain and make other parts of the website unresponsive.

One possible alternative to this is to use HEAD requests to check the availability of a given URL without actually downloading the contents. Currently the VideoContext doesn't do this, but I'll have a look into what it would take to add that functionality.

Kind regards,
Matt

**Comment by Statfine** [comment]
var ctx = new VideoContext();
var videoNode = ctx.createVideoSourceNode('video.mp4');
videoNode.start(5);
videoNode.stop(10);
var videoNode2 = ctx.createVideoSourceNode('video.mp4');
videoNode2.start(20);
videoNode2.stop(25);

I still want to know when there are no content in VideoContext.
Should i take setInterval if i want do something in 10~20s

**Comment by MatthewShotton** [comment]
Hi,

You can use the ```registerTimelineCallback()``` like so:

```
var ctx = new VideoContext();
var videoNode = ctx.createVideoSourceNode('video.mp4');
videoNode.start(5);
videoNode.stop(10);
var videoNode2 = ctx.createVideoSourceNode('video.mp4');
videoNode2.start(20);
videoNode2.stop(25);

ctx.registerTimelineCallback(10, function(){
    console.log("Do something");
});
```

This callback will automatically trigger again if you seek back to the start of the composition, and is closely tied into the timeline of the VideoContext, so will automatically pause if videos start buffering etc.


Kind regards,
Matt

**Comment by Statfine** [comment]
no no, i know registerTimelineCallback.But the time is not sure, in this code is 10s(because 10s~20s there is no content), maybe is begin 10s until 15s, begin 20s until 30s or 50s until over in other code。
The  listen in time is from no content until next source coming.

**Comment by MatthewShotton** [comment]
Ahhh, OK. I think i understand.

Would something like the following fit your requirements?

```
ctx.registerCallback("nocontent", (region)=>{
    console.log("No content form time", region[0], "to time", region[1]);
});
```

**Comment by Statfine** [comment]
bingo, i want to listen when there are no content, then  like change iframe to show no  content in this time or do others

**Comment by MatthewShotton** [comment]
OK, great. I'll look at adding in that functionality.

**Comment by MatthewShotton** [comment]
Hi!

I've pushed update 1e5e07b7a85b7ed346871b9b9b9aa0f0d56bcdcd which is published to npm as v0.34.0 which allows you to register two new VideoContext callbacks.

```
videoContext.registerCallback("content", ()=>{
    console.log("content);
});

videoContext.registerCallback("nocontent", ()=>{
    console.log("no content);
});
```

I haven't added in a time range being passed in to the callbacks because new sourceNodes could be dynamically created and sequenced in a way that would make the time region in a previous callback suddenly invalid.

Hopefully the combination of these two new callbacks will cover your the use-cases?

Kind regards,
Matt

**Comment by Statfine** [comment]
Thanks, It is very useful for me

**Comment by Statfine** [comment]
var ctx = new VideoContext();
var videoNode = ctx.createVideoSourceNode('video.mp4');
videoNode.start(5);
videoNode.stop(10);

ctx.registerCallback("nocontent", ()=>{
    console.log("no content);
});

when the first time play, no console.log during 0s~5s

**Comment by MatthewShotton** [comment]
hi @Statfine,

Thanks for finding this, looks like it was caused by an internal state variable (_sourcesPlaying) being initialised incorrectly so the "nocontent" callback wasn't being fired if it was right at the beginning of a composition.

A fix has been pushed to npm as v0.34.4

Cheers!
